### PR TITLE
Add mass SPICE replacement

### PR DESCRIPTION
**Pull request by martinpitt**
When opening the SPICE dialog from the overview page and there are other
SPICE VMs, offer them for conversion as well, with a menu. Reconfigure
the dialog title and verbiage accordingly.

To make some room in the dialog, move the details of what happens into a popover.

----

[pixel diff review](https://github.com/cockpit-project/pixel-test-reference/compare/e47ae65d70b617aa594aed3964dcd948da668dc4..04230218867ee5dbd413ed03d8474617ec332f70)

This implements @garrett's [remaining design for this feature](https://github.com/cockpit-project/cockpit-machines/pull/1387#issuecomment-1895353645).

## Machines: Mass SPICE conversion

The "Replace SPICE devices" action got introduced in the [previous version](https://cockpit-project.org/blog/cockpit-310.html) to convert SPICE-using VMs to run on a host without SPICE capability. When opened from the overview, that dialog now offers you to convert multiple or all VMs in a single step.

Screen Shot 2024-02-08 at 16 36 00

### CODE REVIEW

File: src/components/vm/vmReplaceSpiceDialog.jsx [hunk 1]
```diff
@@ -32,15 +35,24 @@ import { domainReplaceSpice } from '../../libvirtApi/domain.js';
 
 const _ = cockpit.gettext;
 
-export const ReplaceSpiceDialog = ({ vmName, vmId, connectionName, vmRunning }) => {
+export const ReplaceSpiceDialog = ({ vm, vms }) => {
+    const spiceVMs = vms?.filter(vm => vm.inactiveXML?.hasSpice);
+    // sort spicVMs by name, put the current VM first
+    spiceVMs?.sort((a, b) => (a.name === vm.name) ? -1 : (b.name === vm.name) ? 1 : a.name.localeCompare(b.name));
```
Comment (cockpituous) [inline src/components/vm/vmReplaceSpiceDialog.jsx:41]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-1416-20240201-163147-ce7caf1d-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: src/components/vm/vmReplaceSpiceDialog.jsx [hunk 2]
```diff
@@ -58,9 +70,44 @@
                 .finally(() => setInProgress(false));
     }
 
+    let vmSelect = null;
+
+    if (isMultiple) {
+        const onSelect = (_ev, item) => setSelected(
+            selected.includes(item)
+                ? selected.filter(id => id !== item)
+                : [...selected, item]
```
Comment (cockpituous) [inline src/components/vm/vmReplaceSpiceDialog.jsx:79]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-1416-20240201-163147-ce7caf1d-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: src/components/vm/vmReplaceSpiceDialog.jsx (hunk 1 above)
Comment (cockpituous) [inline src/components/vm/vmReplaceSpiceDialog.jsx:41]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-1416-20240203-041307-a2627c26-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: src/components/vm/vmReplaceSpiceDialog.jsx (hunk 2 above)
Comment (cockpituous) [inline src/components/vm/vmReplaceSpiceDialog.jsx:79]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-1416-20240203-041307-a2627c26-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: src/components/vm/vmReplaceSpiceDialog.jsx (hunk 1 above)
Comment (cockpituous) [inline src/components/vm/vmReplaceSpiceDialog.jsx:41]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-1416-20240203-084715-a2627c26-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: src/components/vm/vmReplaceSpiceDialog.jsx (hunk 2 above)
Comment (cockpituous) [inline src/components/vm/vmReplaceSpiceDialog.jsx:79]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-1416-20240203-084715-a2627c26-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: src/components/vm/vmReplaceSpiceDialog.jsx (hunk 1 above)
Comment (cockpituous) [inline src/components/vm/vmReplaceSpiceDialog.jsx:41]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-1416-20240205-054255-d45075bd-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: src/components/vm/vmReplaceSpiceDialog.jsx (hunk 2 above)
Comment (cockpituous) [inline src/components/vm/vmReplaceSpiceDialog.jsx:79]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-1416-20240205-054255-d45075bd-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: src/components/vm/vmReplaceSpiceDialog.jsx (hunk 1 above)
Comment (cockpituous) [inline src/components/vm/vmReplaceSpiceDialog.jsx:41]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-1416-20240205-065324-d45075bd-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: src/components/vm/vmReplaceSpiceDialog.jsx [hunk 3]
```diff
@@ -58,9 +70,44 @@ export const ReplaceSpiceDialog = ({ vmName, vmId, connectionName, vmRunning })
...
```
Comment (cockpituous) [inline src/components/vm/vmReplaceSpiceDialog.jsx:79]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-1416-20240205-065324-d45075bd-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

**Comment by martinpitt** [comment]
@garrett Could you please do a first review of the screenshot only? Once you are happy, I'll add a pixel test here, but given how fast they get out of date, I'd rather do that as the last step.

**Review by garrett — CHANGES_REQUESTED** [review]
@martinpitt: Apologies. I had the review ongoing for several hours today. I needed to respond with a mockup, and kept getting sidetracked and was in a few meetings since too.

Here's how it should probably be:

image

1. Move the details into a popover.
2. Move the alert into helper text, as it's proactive message instead of a reactive one.

I also had a mockup with the select all as a footer, but it looks a bit odd without a separator. But the menu has a separator with footers, so I think that might work?

image

And do we want "Deselect others" or "Deselect other VMs" instead of "Select none"?

image


(On the topic of "Deselect" vs "Unselect": https://english.stackexchange.com/questions/18465/unselect-or-deselect)

**Comment by martinpitt** [comment]
@garrett : I moved the details to the popover, did the "Deselect others" behaviour/string/layout change, and changed the verbiage:

image

It's now very close, except for turning the alert into a helper text. This is a bit fiddly and needs to happen across all dialogs, so if it's ok with you I'd rather do this in a separate PR?

**Comment by garrett** [comment]
The fix for the blurred line:

```scss
.pf-v5-c-menu.pf-m-scrollable {
    --pf-v5-c-menu__footer--BoxShadow: 0 -1px var(--pf-v5-global--BorderColor--100);
}
```

This changes the blurry shadow (which looks like a bug) for a sharp 1px border.

(We might even want to do this in PF overrides, if we do something similar as to this.)

File: src/components/vm/vmReplaceSpiceDialog.jsx [hunk 4]
```diff
@@ -74,16 +123,21 @@ export const ReplaceSpiceDialog = ({ vmName, vmId, connectionName, vmRunning })
...
-                <Text>{_("Reconfigure the virtual machine for a host which does not support SPICE, for host upgrades or live migration:")}</Text>
+                <Text>{
+                    isMultiple
+                        ? _("Reconfigure the selected virtual machines for a host which does not support SPICE, for host upgrades or live migration:")
+                        : _("Reconfigure the virtual machine for a host which does not support SPICE, for host upgrades or live migration:")
+                }</Text>
                 <TextList>
                     <TextListItem>{_("Convert SPICE graphics console to VNC.")}</TextListItem>
                     <TextListItem>{_("Convert QXL video card to VGA.")}</TextListItem>
                     <TextListItem>{_("Remove SPICE audio and host devices.")}</TextListItem>
```
Comment (garrett) [inline src/components/vm/vmReplaceSpiceDialog.jsx:137]: Nitpick: For lists with fragment phrases, you don't use periods, unless there are two separate thoughts, then you'd use some kind of punctuation to separate them and apply it to all (to stay consistent).

For this list, we shouldn't use periods for any of them.

```suggestion
                    <TextListItem>{_("Convert SPICE graphics console to VNC")}</TextListItem>
                    <TextListItem>{_("Convert QXL video card to VGA")}</TextListItem>
                    <TextListItem>{_("Remove SPICE audio and host devices")}</TextListItem>
```

PatternFly doesn't touch on this, except for headings and titles @ https://www.patternfly.org/ux-writing/punctuation — but it's a general thing when writing, even outside of computers. https://proofed.com/writing-tips/should-bullet-point-lists-end-with-a-period/

It's technically not wrong, as-is, but it's still best to leave out periods for UIs, unless needed.

**Review by garrett — DISMISSED** [review]
Looks great! Thanks.

One very minor thing: The list has periods, and it shouldn't unless needed. (I'm glad it's at least self-consistent!)

File: src/components/vm/vmReplaceSpiceDialog.jsx (hunk 4 above)
Comment (martinpitt) [reply src/components/vm/vmReplaceSpiceDialog.jsx:137]: Done, thanks!

**Comment by martinpitt** [comment]
Popover:

image

I'll update the screenshot in the description for the "closed popover" state.

File: src/components/vm/vmReplaceSpiceDialog.jsx [hunk 5]
```diff
@@ -30,17 +35,28 @@ import { fmt_to_fragments } from 'utils.jsx';
...
 
+import './vmReplaceSpiceDialog.scss';
+
 const _ = cockpit.gettext;
 
-export const ReplaceSpiceDialog = ({ vmName, vmId, connectionName, vmRunning }) => {
+export const ReplaceSpiceDialog = ({ vm, vms }) => {
+    const spiceVMs = vms?.filter(vm => vm.inactiveXML?.hasSpice);
+    // sort spicVMs by name, put the current VM first
+    spiceVMs?.sort((a, b) => (a.name === vm.name) ? -1 : (b.name === vm.name) ? 1 : a.name.localeCompare(b.name));
```
Comment (cockpituous) [inline src/components/vm/vmReplaceSpiceDialog.jsx:45]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-1416-20240208-145413-12ec3af2-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: src/components/vm/vmReplaceSpiceDialog.jsx [hunk 6]
```diff
@@ -58,9 +74,47 @@
                 .finally(() => setInProgress(false));
     }
 
+    let vmSelect = null;
+
+    if (isMultiple) {
+        const onSelect = (_ev, item) => setSelected(
+            selected.includes(item)
+                ? selected.filter(id => id !== item)
+                : [...selected, item]
```
Comment (cockpituous) [inline src/components/vm/vmReplaceSpiceDialog.jsx:83]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-1416-20240208-145413-12ec3af2-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

**Comment by martinpitt** [comment]
Uh, [this is relevant](https://cockpit-logs.us-east-1.linodeobjects.com/pull-1416-20240208-145413-12ec3af2-fedora-39-devel/log.html#85) -- PF doesn't like trying to put a TextList into a Popover?

**Review by garrett — DISMISSED** [review]
Thanks!

File: src/components/vm/vmReplaceSpiceDialog.jsx [hunk 7]
```diff
@@ -30,17 +36,28 @@ import { fmt_to_fragments } from 'utils.jsx';
...
 
+import './vmReplaceSpiceDialog.scss';
+
 const _ = cockpit.gettext;
 
-export const ReplaceSpiceDialog = ({ vmName, vmId, connectionName, vmRunning }) => {
+export const ReplaceSpiceDialog = ({ vm, vms }) => {
+    const spiceVMs = vms?.filter(vm => vm.inactiveXML?.hasSpice);
+    // sort spicVMs by name, put the current VM first
+    spiceVMs?.sort((a, b) => (a.name === vm.name) ? -1 : (b.name === vm.name) ? 1 : a.name.localeCompare(b.name));
```
Comment (cockpituous) [inline src/components/vm/vmReplaceSpiceDialog.jsx:46]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-1416-20240208-153345-081e188a-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: src/components/vm/vmReplaceSpiceDialog.jsx [hunk 8]
```diff
@@ -58,9 +75,45 @@ export const ReplaceSpiceDialog = ({ vmName, vmId, connectionName, vmRunning })
                 .finally(() => setInProgress(false));
     }
 
+    let vmSelect = null;
+
+    if (isMultiple) {
+        const onSelect = (_ev, item) => setSelected(
+            selected.includes(item)
+                ? selected.filter(id => id !== item)
+                : [...selected, item]
```
Comment (cockpituous) [inline src/components/vm/vmReplaceSpiceDialog.jsx:84]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-1416-20240208-153345-081e188a-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: src/components/vm/vmReplaceSpiceDialog.jsx (hunk 7 above)
Comment (cockpituous) [inline src/components/vm/vmReplaceSpiceDialog.jsx:46]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-1416-20240208-171835-081e188a-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: src/components/vm/vmReplaceSpiceDialog.jsx (hunk 8 above)
Comment (cockpituous) [inline src/components/vm/vmReplaceSpiceDialog.jsx:84]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-1416-20240208-171835-081e188a-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: src/components/vm/vmReplaceSpiceDialog.jsx [hunk 9]
```diff
@@ -30,17 +36,28 @@ import { fmt_to_fragments } from 'utils.jsx';
...
+export const ReplaceSpiceDialog = ({ vm, vms }) => {
+    const spiceVMs = vms?.filter(vm => vm.inactiveXML?.hasSpice);
+    // sort spicVMs by name, put the current VM first
+    spiceVMs?.sort((a, b) => (a.name === vm.name) ? -1 : (b.name === vm.name) ? 1 : a.name.localeCompare(b.name));
+    const isMultiple = spiceVMs?.length > 1;
+
     const Dialogs = useDialogs();
     const [error, dialogErrorSet] = useState(null);
     const [inProgress, setInProgress] = useState(false);
+    const [selected, setSelected] = useState(isMultiple ? [spiceVMs[0].id] : []);
```
Comment (mvollmer) [inline src/components/vm/vmReplaceSpiceDialog.jsx:52]: It's quite confusing how the "primary" VM (passed in `vm`) is handled re the various lists of vms. The primary is part of `spiceVms`, and might be part of `selected` (but only if `isMultiple` is true), so it might also be part of `selectedVMs` (in the multi case), and wouldn't it then be converted twice?

File: src/components/vm/vmReplaceSpiceDialog.jsx [hunk 10]
```diff
@@ -30,17 +36,28 @@ import { fmt_to_fragments } from 'utils.jsx';
...
 
     function onReplace() {
         setInProgress(true);
 
+        const selectedVMs = spiceVMs?.filter(vm => selected.includes(vm.id)) || [];
+
         return domainReplaceSpice({ connectionName: vm.connectionName, id: vm.id })
```
Comment (mvollmer) [inline src/components/vm/vmReplaceSpiceDialog.jsx:59]: So the primary VM always gets replaced, eventhough there is a checkbox for it in the multi case that the user might uncheck?

File: src/components/vm/vmReplaceSpiceDialog.jsx [hunk 11]
```diff
@@ -30,17 +36,28 @@ import { fmt_to_fragments } from 'utils.jsx';
...
+                .then(() => Promise.allSettled(selectedVMs.map(vm => domainReplaceSpice({ connectionName: vm.connectionName, id: vm.id }))))
```
Comment (mvollmer) [inline src/components/vm/vmReplaceSpiceDialog.jsx:60]: This converts all VMs in parallel, which is fine for Cockpit and pipelining over slow connections, but might be stressing poor libvirt. Would it make sense to be conservative and convert them one-by-one? Or do we know that libvirt is fine with this?

**Review by mvollmer — COMMENTED** [review]
The code was a bit hard to follow, and I think the primary VM will be replaced twice (without any ill effect, probably). Please check that and fix if you think it is a good idea.

File: src/components/vm/vmReplaceSpiceDialog.jsx (hunk 10 above)
Comment (martinpitt) [reply src/components/vm/vmReplaceSpiceDialog.jsx:59]: Indeed, thanks for spotting! I created an integration test for covering this case. The dialog is rather complicated due to the possibility of unchecking the VM which you triggered the action on :cry:

File: src/components/vm/vmReplaceSpiceDialog.jsx (hunk 9 above)
Comment (martinpitt) [reply src/components/vm/vmReplaceSpiceDialog.jsx:52]: Right, the dialog needs to handle both cases (single and multi-VM) separately. Thanks for spotting!

File: src/components/vm/vmReplaceSpiceDialog.jsx (hunk 11 above)
Comment (martinpitt) [reply src/components/vm/vmReplaceSpiceDialog.jsx:60]: The conversion is not that expensive, but let's check. My VM can host about 60 running micro-VMs until it runs out of memory:
```
for i in `seq 60`; do virt-install --memory 50 --pxe --virt-type qemu --os-variant alpinelinux3.8 --disk none --wait 0 --name test$i; done
```
This takes about 5 seconds and succeeds. I'll try again with serial conversion, if that doesn't take an undue amount of time, we can play it safe and do it serially.

File: src/components/vm/vmReplaceSpiceDialog.jsx [hunk 12]
```diff
@@ -30,42 +36,95 @@ import { fmt_to_fragments } from 'utils.jsx';
...
+        }
+    }
+
+    let vmSelect = null;
+
+    if (isMultiple) {
+        const onSelect = (_ev, item) => setSelected(
+            selected.includes(item)
+                ? selected.filter(id => id !== item)
+                : [...selected, item]
```
Comment (cockpituous) [inline src/components/vm/vmReplaceSpiceDialog.jsx:89]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-1416-20240209-102132-d1b11d6f-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: src/components/vm/vmReplaceSpiceDialog.jsx (hunk 11 above)
Comment (mvollmer) [reply src/components/vm/vmReplaceSpiceDialog.jsx]: > This takes about 5 seconds and succeeds.

The conversion takes 5 seconds, right? (Not the creation of the VMs.)

Comment (martinpitt) [reply src/components/vm/vmReplaceSpiceDialog.jsx]: Right, I meant the "Replace" button. Creating the VMs takes much longer.

File: src/components/vm/vmReplaceSpiceDialog.jsx [hunk 13]
```diff
@@ -30,42 +36,95 @@ import { fmt_to_fragments } from 'utils.jsx';
...
 
+import './vmReplaceSpiceDialog.scss';
+
 const _ = cockpit.gettext;
 
-export const ReplaceSpiceDialog = ({ vmName, vmId, connectionName, vmRunning }) => {
+export const ReplaceSpiceDialog = ({ vm, vms }) => {
+    const spiceVMs = vms?.filter(vm => vm.inactiveXML?.hasSpice);
+    // sort spiceVMs by name, put the current VM first
+    spiceVMs?.sort((a, b) => (a.name === vm.name) ? -1 : (b.name === vm.name) ? 1 : a.name.localeCompare(b.name));
```
[the remark this answers is not in the document]
Comment (martinpitt) [reply src/components/vm/vmReplaceSpiceDialog.jsx:46]: They are guaranteed *not* to be unique across different connections; you can easily have a "test1" on system and session. Indeed this could be `a === vm` and `b === vm`, I'll test if that works.

File: src/components/vm/vmReplaceSpiceDialog.jsx [hunk 14]
```diff
@@ -30,42 +36,95 @@ import { fmt_to_fragments } from 'utils.jsx';
...
+    const isMultiple = spiceVMs?.length > 1;
+
     const Dialogs = useDialogs();
     const [error, dialogErrorSet] = useState(null);
     const [inProgress, setInProgress] = useState(false);
+    const [selected, setSelected] = useState(isMultiple ? [spiceVMs[0].id] : []);
```
[the remark this answers is not in the document]
Comment (martinpitt) [reply src/components/vm/vmReplaceSpiceDialog.jsx:52]: That's indeed what I had first, but it doesn't work: `selected` contains items which must be representable in the DOM.But I'll prepend `vm.connectionName` to make them unique.

File: src/components/vm/vmReplaceSpiceDialog.jsx [hunk 15]
```diff
@@ -30,42 +36,98 @@ import { fmt_to_fragments } from 'utils.jsx';
...
+        }
+    }
+
+    let vmSelect = null;
+
+    if (isMultiple) {
+        const onSelect = (_ev, item) => setSelected(
+            selected.includes(item)
+                ? selected.filter(id => id !== item)
+                : [...selected, item]
```
Comment (cockpituous) [inline src/components/vm/vmReplaceSpiceDialog.jsx:92]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-1416-20240209-140217-7035aafd-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

**Review by mvollmer — APPROVED** [review]
Danke!
